webapp/src/dogma/features/settings/mirror-access/MirrorAccessControlForm.tsx

```diff
@@ -0,0 +1,201 @@
+  Spacer,
+  Stack,
+  VStack,
+} from '@chakra-ui/react';
+import { HiOutlineIdentification } from 'react-icons/hi';
+import { Controller, useForm } from 'react-hook-form';
+import React, { useMemo } from 'react';
+import { LabelledIcon } from 'dogma/common/components/LabelledIcon';
+import FieldErrorMessage from 'dogma/common/components/form/FieldErrorMessage';
+import { MirrorAccessControlRequest } from 'dogma/features/settings/mirror-access/MirrorAccessControl';
+import { LuRegex } from 'react-icons/lu';
+import { MdOutlineDescription, MdPolicy } from 'react-icons/md';
+import { RiSortNumberAsc } from 'react-icons/ri';
+
+interface MirrorAccessControlFormProps {
+  defaultValue: MirrorAccessControlRequest;
+  onSubmit: (credential: MirrorAccessControlRequest, onSuccess: () => void) => Promise<void>;
+  isWaitingResponse: boolean;
+}
+
+const MirrorAccessControlForm = ({
+  defaultValue,
+  onSubmit,
+  isWaitingResponse,
+}: MirrorAccessControlFormProps) => {
+  const isNew = defaultValue.id === '';
+  const {
+    register,
+    handleSubmit,
+    setValue,
+    control,
+    formState: { errors, isDirty },
+  } = useForm<MirrorAccessControlRequest>({
+    defaultValues: {},
+  });
+
+  useMemo(() => {
+    if (!isNew) {
+      // @ts-expect-error 'allow' in a radio group is a string
+      setValue('allow', defaultValue.allow + '');
+    }
+  }, [isNew, setValue, defaultValue.allow]);
+
+  return (
+    <form onSubmit={handleSubmit((data) => onSubmit(data, () => {}))}>
+      <Center>
+        <VStack width="80%" align="left">
+          <Heading size="lg" mb={4}>
+            {isNew ? 'New Mirror Access Control' : 'Edit Mirror Access Control'}
+          </Heading>
+          <FormControl isRequired isInvalid={errors.id != null}>
+            <FormLabel>
+              <LabelledIcon icon={HiOutlineIdentification} text="ID" />
+            </FormLabel>
+            <Input
+              id="id"
+              name="id"
+              type="text"
+              readOnly={!isNew}
+              placeholder="ID"
+              defaultValue={defaultValue.id}
+              {...register('id', { required: true, pattern: /^[a-zA-Z0-9-_.]+$/ })}
+            />
+            {errors.id ? (
+              <FieldErrorMessage error={errors.id} fieldName="ID" />
+            ) : (
+              <FormHelperText>
+                The mirror access control ID must be unique and contain alphanumeric characters, dashes,
+                underscores, and periods only.
+              </FormHelperText>
+            )}
+          </FormControl>
+          <Spacer />
+
+          <FormControl isRequired isInvalid={errors.targetPattern != null}>
+            <FormLabel>
+              <LabelledIcon icon={LuRegex} text="Git URI pattern" />
+            </FormLabel>
+            <Input
+              id="targetPattern"
+              name="targetPattern"
+              placeholder=".*://github.com/line/.*"
+              defaultValue={defaultValue.targetPattern}
```

Comment:
Can we validate the pattern here as well?

Reply:
`targetPattern` format is not fixed. It could be a plain text or a regular expression.
Do you intend to check if `targetPattern` is a valid regex?

Reply:
NVM. I missed the point that it can be plaintext.